crates/bevy_reflect/src/struct_trait.rs

```diff
@@ -69,32 +69,35 @@ pub trait Struct: Reflect {
 /// A container for compile-time struct info.
 #[derive(Clone, Debug)]
 pub struct StructInfo {
+    name: &'static str,
     type_name: &'static str,
     type_id: TypeId,
     fields: Box<[NamedField]>,
-    field_indices: HashMap<Cow<'static, str>, usize>,
+    field_names: Box<[&'static str]>,
```

Comment:
> That being said, I think I understand why you added these here: the serde `deserializer.deserialize_struct` function requires `&'static [&'static str]` field names.

Yep that's exactly why I added them haha

> However! This is just a "hint" (both according to the docs and from my own experience). If the deserialized file does not contain a field in the list, it will just be ignored during deserialization. If the passed in list does not contain a field that exists in the file, that field will still be visited! I experimented by passing in `&[]` (which is implicitly static) instead of the actual field names and everything still works!

Yeah I just tested out a couple other common deserializers (JSON and TOML), and it does seem to work fine if we just pass it `&[]`. I'll remove `field_names` and their usage for now. If any deserializer comes up that really needs them, we can always revert that change

Reply:
On the other hand, though, doing this allows us to provide better errors. We can easily do something like:

```rust
let field = struct_info
  .get_field(&ident)
  .ok_or_else(|| serde::de::Error::unknown_field(&ident, struct_info.get_field_names()))?;
```

Which gives us something like:

```
Error { code: Message("unknown field `baz`, expected one of `foo`, `bar`")
```

Is this worth keeping them around? Or is it still not worth keeping the static array?

Reply:
> If any deserializer comes up that really needs them, we can always revert that change

Sounds good. I'm guessing/hoping this should work as expected everywhere, given that it is described as a hint. But even if we find one that doesn't work, from a Bevy perspective I don't think thats particularly important. Of course bevy_reflect is intended to be a generic reflection system (and that does include serialization), but Bevy's needs come first (and I think that includes optimizing this case over supporting everything).

> Is this worth keeping them around? Or is it still not worth keeping the static array?

We don't need static arrays for that. We can just do `struct_info.iter().map(|field| field.name())`.

Reply:
Ahh `serde::de::Error::unknown_field` _does_ require a slice of &'static strs.  Not a huge loss. We can just use `serde::de::Error::custom` and achieve effectively the same thing.

Reply:
Updated! We now produce errors that match the ones produced by serde.